### DISCUSSION TITLE
Tuning for large compile task

### DISCUSSION THREAD

**Discussion by myc724** [juicedata/juicefs#686] [Q&A] [answered]
We’ve deployed JuiceFS (version 0.11.0-37 2021-03-19 6127863) on the top of two Redis 6.2.2 (one for metadata and the one for object-storage) using “--io-threads 16" and “-io-threads-do-reads yes” but found that, for the following application, JuiceFS lagging behind a standard NFS (tier2) storage. After some tuning exercises, all knob tweaking (eg. smaller block-size, larger cache-size) seems only result incremental speedup. Our next step is to try Redis cluster (with at least 3 Sentinals to guarantee the atomicity of metadata operations) or TiKV. Is there any obvious thing that we can try before deep dive into distributed database solution? Thanks!

The application is to compile 140GB source code (1.2M files - avg. 113K per file) using 1250 CPUs and wrote 70GB object files (115K files - avg. 608KB per obj).

- NFS: 80 minutes.
- JuiceFS: 235 minutes.
- Two JuiceFS each has own servers (one for reading source code the other one for writing object): 200 minutes.

By switching source-code and object to various partitions and exhausting all combinations, we noticed that reading the 140GB source code is the bottleneck. Also, Redis config file seems implies that multiple threading might NOT help reading so much, so we’re looking into distributed database solutions. Please let us know if any obvious tuning that we can try. Thanks!

**Comment by xiaogaozi** [comment, marked as answer]
Hi @myc724, thanks for your feedback.

First of all, the v0.11.0 of JuiceFS is too old (latest version is v0.15.2), could you try the latest version?

> we noticed that reading the 140GB source code is the bottleneck

You mean read the metadata or data? JuiceFS has a `profile` subcommand could help you profiling your case, please refer to [document](https://github.com/juicedata/juicefs/blob/main/docs/en/operations_profiling.md) for more details.

> Our next step is to try Redis cluster (with at least 3 Sentinals to guarantee the atomicity of metadata operations) or TiKV

If my understanding is correct, the "Redis cluster" you mentioned is [this one](https://redis.io/topics/cluster-tutorial). But JuiceFS relies on Redis transaction to guarantee atomicity, Sentinel seems helpless in this case?

TiKV metadata engine is ready now in `main` branch, you could try it immediately, please refer to [document](https://github.com/juicedata/juicefs/blob/main/docs/en/databases_for_metadata.md#tikv) for more details.

**Reply by myc724** [reply]
Thanks, xiaogaozi! We are going to
1. deploy the latest JuiceFS,
2. use the profile command to identify the bottleneck (... our guess is metadata ... will see), as well as
3. explore TiKV for distributed key-value DB solution.

**Reply by davies** [reply]
There are a few things to speed up:
1. Use the latest JuiceFS, which will enable page cache for repeated reading on same file.
2. Increase the --cache-size to improve the cache hit rate for data. You can setup the p8s and grafana the check the metrics.
3. Increase --attr-cache --entry-cache --dir-entry-cache to reduce the lookup call into JuiceFS.
4. Mount JuiceFS with '-o writeback_cache' to enable writeback in kernel (require JuiceFS 0.15+ and kernel 3.15+). This can improve the performance writing objects.

Can we use the linux kernel or other large open source project to simulate your situation?

**Reply by davies** [reply]
Also you could enable `--writeback` to speed up writing small files (after https://github.com/juicedata/juicefs/pull/696 merged into main branch).

**Reply by myc724** [reply]
Thanks a lot, Davies! We will try these switches in the next tunning activities. Let me know think of how you can simulate our situation.

**Reply by myc724** [reply]
Another quick question: can "--read-only" benefit reading performance? Thanks!

In detail, we could use "--read-only" to create the mount-point only for reading source code (during distributed compile) then set up another volume/mount-point dedicated for object writing. Just wonder whether a "--read-only" mount-point has any performance benefit comparing to regular (ie. read-write) mount-point.

**Reply by davies** [reply]
A  "--read-only" mount-point should still see changes from other client, so it will not do aggressive cache to boost the performance.

**Reply by myc724** [reply]
I see. Thanks! So, we should NOT use "--read-only" to attempt to gain some performance benefit, but we can use "--read-only" for those mount points we would like to restrict the write access.

**Reply by myc724** [reply]
Just to get an idea of how increased the value of these three options can reduce the lookup call into JuiceFS. Please let us know if the statement is incorrect. Thanks!

Are these the timeout for caching operations? If so, after increasing these three timeouts to 10 seconds, JuiceFS will cache “more” key-value pairs for the attributes/file-entries/ dir-entries being used.
- --attr-cache value ## attributes cache timeout in seconds (default: 1)
- --entry-cache value ## file entry cache timeout in seconds (default: 1)
- --dir-entry-cache value ## dir entry cache timeout in seconds (default: 1)

**Reply by davies** [reply]
Yes, these are the timeout to keep the cache in kernel. Once you are the only one to access these source code, it's safe to increase them much higher.

**Reply by myc724** [reply]
Thanks! We're going to start with 10 seconds and will jump to minute(s) if it shows improvement.

**Reply by SandyXSD** [reply]
@myc724 Hi, I tried some tuning on the Redis project for test, and observes that:

- For git related commands, e.g. `git status`, which emit a lot of `lookup` requests, increasing both `--attr-cache` and `--entry-cache` at the same time is helpful, since the requests can be handled by kernel and replied immediately

- For compiling process, writing to object storage becomes the bottleneck, so enabling `--writeback` and set `--cache-dir` to a fast local disk will help a lot.

Followings are time costs to `make` the Redis project:

1. Local SSD: 0m29.348s
2. JuiceFS with Redis & S3, default configuration: 2m47.335s
3. 2 + `--attr-cache 60 --entry-cache 60`: 2m41.040s
4. 2 + `--writeback`: 0m38.308s

Also we are developing a tool (#665 ) to show more internal statistics, which should make diagnosing easier.

**Reply by myc724** [reply]
Thanks for the detailed explanation! We will look into deploying these options.

**Reply by myc724** [reply]
Further questions. In some cases, we’re seeing the following errors from juicefs clients syslog, while the server is still running but somewhat busy (but not thrashing). We’re looking into TiKV now. Meanwhile, besides boosting metadata server performance, are there any client side option that we can use to get over these IO errors (eg. for flushing/uploading-chunk)? Thanks!

2021/08/04 17:26:35.222938 juicefs[23063] <ERROR>: flush 152655 timeout after waited 5m0s
2021/08/04 17:26:35.223004 juicefs[23063] <ERROR>: pending slice 152655-0: {id:10465721 chunk:0xc006fcc270 off:0 length:12581280 soff:0 slen:12581280 writer:0xc0000ab9e0 freezed:true done:false err:0 notify:0xc001c8ac90 started:{wall:13851640171213559405 ext:10724743684966 loc:0x41292e0} lastMod:{wall:13851640171333245385 ext:10724789629131 loc:0x41292e0}}
2021/08/04 17:26:35.223757 juicefs[23063] <WARNING>: All goroutines (45):
goroutine 800080 [running]:
github.com/juicedata/juicefs/pkg/vfs.(*fileWriter).flush(0xc0025b0280, 0x7fffcc287a40, 0xc00175a0c0, 0x415af00, 0x0)
  /u/tools/juicefs/pkg/vfs/writer.go:372 +0x5f2
…

2021/08/04 17:27:14.165449 juicefs[23063] <ERROR>: upload chunk 10465721 (length: 12581280) fail: upload block chunks/10/10465/10465721_2_4192672: write tcp 10.15.161.8:35748->10.15.24.254:1225: i/o timeout (after 11 tries)
2021/08/04 17:27:26.236417 juicefs[23063] <INFO>: slow request: DELETE chunks/10/10465/10465721_0_4194304 (read tcp 10.15.161.8:35770->10.15.24.254:1225: i/o timeout, 12.071s)
2021/08/04 17:27:28.362639 juicefs[23063] <WARNING>: write inode:152655 error: input/output error
2021/08/04 17:27:28.362687 juicefs[23063] <ERROR>: write inode:152655 indx:0  input/output error
…

**Reply by davies** [reply]
Is 10.15.24.254:1225 the address of your Redis instance?

It seems there is a performance issue on the Redis who stores data. How much memory it has? Is swap enabled in that machine?

**Reply by myc724** [reply]
Yes, it was the Redis server that I used for the object-store. The server has swap enabled. The memory consumption was around 70GB. But, the host machine has 3TB RAM (and 244 cores) and the memory usage was not near the threshold at all. Also, I ran Redis with 32 IO threads (read-write enabled), but "top" DID NOT show the system started thrashing at all. I guess ... if IO error came from 5 minutes timeout, I can only boost server performance. We're looking into TiKV now.

**Reply by davies** [reply]
Maybe Redis (for meta) + TiKV (for data) could be good choice.

**Reply by myc724** [reply]
Does "write inode:152655 error: input/output error" indicate the metadata server was also stressed (since recording inode should be handled by metadata server)?

**Reply by davies** [reply]
no, that means it failed to write data into Redis
